Added Dataset.split. Fix incorrect version. Move tool functions to module

samedii opened the pull request:
- Added dataset.split
- Fixed incorrect version (was using ml-workflow...)
- Move tool functions to tools module so they are less exposed to the user in the API

Jim-Holmstroem commented on datastream/dataset.py:
```diff
@@ -358,3 +434,38 @@ def test_combine_dataset():
+        stratify=np.concatenate([np.ones(50), np.zeros(50)]),
+    ))).map(tuple)
+
+    split_file = Path('test_split_dataset.json')
```
you can use the `io.StringIO` to avoid using files
https://docs.python.org/3/library/io.html

samedii replied:
Yes I figured but didn't want to put time into learning how to yet. Also I don't think StringIO will be easy to understand in the documentation so would be nice to hide the code or something for the user. That's probably possible as well but also something I didn't want to put time into learning (yet at least).

I think I saw some pytest solution to mocking files. Maybe that's a better solution

samedii replied:
Looked into StringIO. It does not work here because we take a pathlib.Path and not a file object. Haven't looked into mocking files, that would probably work. I don't think we have to change this now.

Jim-Holmstroem commented on datastream/dataset.py:
```diff
@@ -156,6 +172,58 @@ def subset(
+        self,
+        key_column: str,
+        proportions: Dict[str, float],
+        filepath: Optional[Union[str, Path]] = None,
```
needs an explanation on what happens if the 3 cases
1. not set
2. set but not exist
3. set and exists

samedii replied:
> Optionally you can chose to stratify on a column in the source dataframe or save the split to a json file.

I think it is explained but I'm biased here.

I think the name `filepath` forces you to look at the documentation though so maybe it should be called `save_filepath` or something?

Jim-Holmstroem replied:
will the split append or overwrite I think is the big thing you need to explain
its clear if you continue reading, but it should be clear directly, it was the first question I had

Jim-Holmstroem commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+        split = json.read(filepath)
+
+        if set(proportions.keys()) != set(split.keys()):
+            raise AssertionError(
```
https://docs.python.org/3/library/exceptions.html#ValueError

FelixAbrahamsson replied:
To add to this we might want to also print what values were received.

samedii replied:
> It's not immediately obvious to the user how to use `Dataset.split` on a dataset created with `Dataset.from_subscriptable`. Do we want to support this? Different proposals:
> 
> * Change `Dataset.from_subscriptable` dataframe key from `'example'` to `'key'`, have default key name in `Dataset.split` be `'key'`
> * Let the user input (optionally) a column name for the dataframe created by `Dataset.from_subscriptable` (I prefer this)

We do not support Dataset.from_subscriptable for this. We might want to make a note of that. Dataset.subset is also not available for datasets created with from_subscriptable.

I prefer reducing the number of ways to input data over trying to support multiple types of input. It becomes messy and buggy very quickly. I'd rather drop Dataset.from_subscriptable and have a clean interface over trying to support many types of input arguments in all our functions.

_Note that Dataset.from_subscriptable basically does nothing. It just wraps the input list in a pandas.DataFrame_

> Create ``Dataset`` based on subscriptable i.e. implements ``__get_item__``. Mostly useful for simple examples.

In the Dataset.from_subscriptable we say that it's mostly useful for simple examples but we could be more explicit. I'll note these limitations so it's known until we might make a bigger decision.

Jim-Holmstroem commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+    - Adapt after removing examples from dataset
+    - Adapt to new stratification
+    '''
+    if abs(sum(proportions.values()) - 1.0) >= 1e-6:
```
how sensitive is this 1e-6?
https://stackoverflow.com/a/19141711/905596
if you close to 10 proportions and are running on 32 bit system you might stumble on problems

FelixAbrahamsson replied:
I think 10 proportions is an unrealistic use-case? Then again, I suppose you could use this to set up k-folds so maybe not? Anyway, it seems unlikely to me that someone would be doing deep learning on a 32-bit system, no?

samedii replied:
We can lower 1e-6 to 1e-5

Jim-Holmstroem commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+        unassigned, size=k, replace=False
+    ).tolist()  
+
+
```
add some degenerate cases like split an empty dataframe, missing stratify column and see if it behaves as intended

samedii replied:
I think Dataset.split needs to be tested a lot more than it has. However, I'm not sure this should be done with pytest and I think we need to discuss what our test coverage goals are. It is all over the place at the moment and I'm not convinced we want to go all out with pytest

samedii replied:
Added some degenerate tests. Did not find any bugs but maybe it will stop future bugs

FelixAbrahamsson commented on datastream/json.py:
```diff
@@ -0,0 +1,11 @@
+import json
```
Perhaps this belongs in `tools`?

samedii replied:
I thought about this too but thought it might be ugly but maybe it's fine. I'll change it

FelixAbrahamsson commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+def stratas(dataframe, stratify_column):
+    return [
+        strata_value for strata_value, _ in (
+            dataframe[stratify_column].value_counts().iteritems()
```
Alternatively
```python 
return [strata for strata in dataframe[stratify_column].unique()]
```

FelixAbrahamsson replied:
Could also have this function return the stratas directly instead of just the strata values?
```
return [
    dataframe[dataframe[stratify_column] == strata_value]
    for strata_value in dataframe[stratify_column].unique()
]
```

samedii replied:
Just copied from the old code. I found a bug thanks to this. Thanks! :)

FelixAbrahamsson commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+        else:
+            for strata in stratas(dataframe, stratify_column):
+                split = split_proportion(
+                    dataframe[key_column],
```
Should be `dataframe[dataframe[stratify_column] == strata]` or am I mistaken?

samedii replied:
Yes! This is the bug I found when I was resolving your comment above :+1:

FelixAbrahamsson commented on datastream/tools/split_dataframe.py:
```diff
@@ -0,0 +1,155 @@
+
+    probability = float_target_split - int(float_target_split)
+    if probability >= 1e-6 and np.random.rand() <= probability:
+        return int(float_target_split) + 1
```
So if I understand correctly this is a probability-based rounding?

samedii replied:
Yes. This is good behaviour if our stratas are too small at least they will still divide the data by the correct ratio

FelixAbrahamsson replied:
I like it :) Just never seen it before.

FelixAbrahamsson commented:
It's not immediately obvious to the user how to use `Dataset.split` on a dataset created with `Dataset.from_subscriptable`. Do we want to support this? Different proposals:
* Change `Dataset.from_subscriptable` dataframe key from `example` to `key`, have default key name in `Dataset.split` be `'key'`
* Let the user input (optionally) a column name for the dataframe created by `Dataset.from_subscriptable` (I prefer this)

samedii commented:
> It's not immediately obvious to the user how to use `Dataset.split` on a dataset created with `Dataset.from_subscriptable`. Do we want to support this? Different proposals:
> 
> * Change `Dataset.from_subscriptable` dataframe key from `'example'` to `'key'`, have default key name in `Dataset.split` be `'key'`
> * Let the user input (optionally) a column name for the dataframe created by `Dataset.from_subscriptable` (I prefer this)

We do not support Dataset.from_subscriptable for this. We might want to make a note of that. Dataset.subset is also not available for datasets created with from_subscriptable.

I prefer reducing the number of ways to input data over trying to support multiple types of input. It becomes messy and buggy very quickly. I'd rather drop Dataset.from_subscriptable and have a clean interface over trying to support many types of input arguments in all our functions.

_Note that Dataset.from_subscriptable basically does nothing. It just wraps the input list in a pandas.DataFrame_

> Create ``Dataset`` based on subscriptable i.e. implements ``__get_item__``. Mostly useful for simple examples.

In the Dataset.from_subscriptable we say that it's mostly useful for simple examples but we could be more explicit. I'll note these limitations so it's known until we might make a bigger decision.

FelixAbrahamsson commented:
Okay, if we mention it in the docs I think that's sufficient. Like you said, from_subscriptable doesnt do much and is just a quick way to create a Dataset. If the user wants access to the other methods they can easily just create a dataframe to input.

samedii commented:
I think most of the things are resolved now. I would like to merge so we can get the embarrassing bug out of the way :D We can get back to mocking the file system in another PR. Thanks for the good points! @FelixAbrahamsson @Jim-Holmstroem :) :+1:
